dynamic_programming/abbreviation.cpp

```diff
@@ -0,0 +1,186 @@
+/**
+ * @file
+ * @brief Implementation of Abbrievation
+ * (https://www.hackerrank.com/challenges/abbr/problem)
+ *
+ * @details
+ * Given two strings, a and b, determine if it's possible to make a equal to
+ * b You can perform the following operations on the string a:
+ * 1. Capitalize zero or more of a's lowercase letters.
+ * 2. Delete all of the remaining lowercase letters in a.
+ *
+ * ##Algorithm
+ * The idea is in the problem statement itself: iterate through characters of
+ * string a and b (for character indexes i and j respectively):
+ * 1. If a[i] and b[j] are equal, then move to next position
+ * 2. If a[i] is lowercase of b[j], then explore two possibilities:
+ * a. Capitalize or
+ * b. Skip a[i]
+ * 3. If the a[i] is not uppercase, just discard the char, else return false
+ *
+ * Time Complexity (O(|a|*|b|)) where |a| => length of string
+ * @author Ashish Daulatabad (https://github.com/AshishYUO)
+ */
+
+#include <cassert>   /// assert function
+#include <iostream>  /// input/output operations
+#include <string>    /// std::string library
+#include <vector>    /// std::vector STL library
+/**
+ * @namespace dynamic_programming
+ * @brief Dynamic Programming Algorithms
+ */
+namespace dynamic_programming {
+/**
+ * @namespace abbreviation
+ * @brief Implementation of Abbreivation problem
+ */
+namespace abbreviation {
+/**
+ * @brief
+ * Recursive Dynamic Programming function
+ * @details
+ * Returns whether `s` can be converted to `t` with following rules:
+ * a. Capitalize zero or more of a's lowercase letters from string `s`
+ * b. remove all other lowercase letters from string `s`
+ * @param memo memo as parameter to store the result
+ * @param visited visited boolean to check if the result is already computed
+ * @param str given string, which might not be abbreivated
+ * @param result resultant abbreivated string
+ * @param str_idx pointer for string `str`, helpful for transitions
+ * @param result_idx pointer for string `result`, helpful for transitions
+ * @returns boolean (`true` or `false`) whether string `str` can be converted to
```

Comment:
Complete the steps below. Fill in the documentation below.
```suggestion
 * @returns `false` if ....
 * @returns `true` if ....
```